### PR TITLE
sched/msgq: add support of System V message queue

### PR DESCRIPTION
**Pull request by anchao**
## Summary

sched/msgq: add support of System V message queue
sched/mqueue: decoupling condition member to common prologue

https://man.openbsd.org/msgget.2

## Impact

N/A

## Testing

System V message queue test

### CODE REVIEW

File: sched/mqueue/msginternal.c [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+/****************************************************************************
+ * Name: nxmsg_lookup
+ ****************************************************************************/
+
+FAR struct msgq_s *nxmsg_lookup(key_t key)
+{
+  FAR struct msgq_s *msgq;
+
+  list_for_every_entry(&g_msgqactivelist, msgq, struct msgq_s, node)
```
Comment (xiaoxiang781216) [inline sched/mqueue/msginternal.c:128]: can we change g_msgqactivelist to an array and identify the msq by index to improve the speed?

File: include/nuttx/mqueue.h [hunk 2]
```diff
@@ -83,16 +83,23 @@
  * Public Type Declarations
  ****************************************************************************/
 
+/* Common prologue of all message queue structures. */
+
+struct mqueue_comm_s
```
[the remark this answers is not in the document]
Comment (anchao) [reply include/nuttx/mqueue.h:88]: Done

File: sched/mqueue/msgget.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  if (msgq)
+    {
+      if ((msgflg & IPC_CREAT) && (msgflg & IPC_EXCL))
+        {
+          ret = -EEXIST;
+        }
+    }
+  else
+    {
+      if (key != IPC_PRIVATE && !(msgflg & IPC_CREAT))
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msgget.c:72]: Done

File: sched/mqueue/msginternal.c (hunk 1 above)
Comment (anchao) [reply sched/mqueue/msginternal.c]: Done

File: fs/procfs/fs_procfsproc.c [hunk 4]
```diff
@@ -411,7 +411,7 @@ static FAR const char * const g_statenames[] =
   "Inactive",
   "Waiting,Semaphore",
   "Waiting,Signal"
-#ifndef CONFIG_DISABLE_MQUEUE
+#if !defined(CONFIG_DISABLE_MQUEUE) && !defined(CONFIG_DISABLE_MQUEUE_SYSV)
```
[the remark this answers is not in the document]
Comment (anchao) [reply fs/procfs/fs_procfsproc.c:414]: Done

File: include/nuttx/mqueue.h [hunk 5]
```diff
@@ -83,16 +83,23 @@
...
+{
+  int16_t nwaitnotfull;       /* Number tasks waiting for not full */
+  int16_t nwaitnotempty;      /* Number tasks waiting for not empty */
+};
+
 /* This structure defines a message queue */
 
 struct mqueue_inode_s
 {
+  struct mqueue_comm_s comm;  /* Common prologue */
```
[the remark this answers is not in the document]
Comment (anchao) [reply include/nuttx/mqueue.h]: Done

File: include/nuttx/mqueue.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply include/nuttx/mqueue.h:97]: no change?

File: include/nuttx/mqueue.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply include/nuttx/mqueue.h]: no change?

**Review by jerpelea — APPROVED** [review]

File: include/nuttx/mqueue.h [hunk 6]
```diff
@@ -96,6 +96,8 @@
 
 struct mqueue_cmn_s
 {
+  dq_queue_t waitfornotempty; /* Task list waiting for not empty */
+  dq_queue_t waitfornotfull;  /* Task list waiting for not full */
```
Comment (xiaoxiang781216) [inline include/nuttx/mqueue.h:100]: Move the related change to the previous patch too

File: sched/mqueue/msginternal.c [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MQ_PERBLOCK 10
+
+/****************************************************************************
+ * Private Type Definitions
+ ****************************************************************************/
+
+struct nxmsglist
```
Comment (xiaoxiang781216) [inline sched/mqueue/msginternal.c:38]: change to other name since we don't use list now

Comment (anchao) [reply sched/mqueue/msginternal.c:38]: Done

File: include/sys/msg.h [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+ *            message queue description referred to by mqdes.
+ *   EINVAL   Either msg or mqdes is NULL or the value of prio is invalid.
+ *   EPERM    Message queue opened not opened for writing.
+ *   EMSGSIZE 'msglen' was greater than the maxmsgsize attribute of the
+ *            message queue.
+ *   EINTR    The call was interrupted by a signal handler.
+ *
+ ****************************************************************************/
+
+int msgsnd(int msqid, const void *msgp, size_t msgsz, int msgflg);
```
Comment (pkarashchenko) [inline include/sys/msg.h:179]:
```suggestion
int msgsnd(int msqid, FAR const void *msgp, size_t msgsz, int msgflg);
```

File: include/sys/msg.h [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+ *
+ * Returned Value:
+ *   On success, msgrcv() returns the number of bytes actually copied
+ *   into the mtext array.
+ *   On failure, both functions return -1, and set errno to indicate
+ *   the error.
+ *
+ ****************************************************************************/
+
+ssize_t msgrcv(int msqid, void *msgp, size_t msgsz, long msgtyp, int msgflg);
```
Comment (pkarashchenko) [inline include/sys/msg.h:205]:
```suggestion
ssize_t msgrcv(int msqid, FAR void *msgp, size_t msgsz, long msgtyp, int msgflg);
```

File: include/nuttx/mqueue.h [hunk 10]
```diff
@@ -85,25 +85,32 @@
...
+  int16_t nwaitnotfull;       /* Number tasks waiting for not full */
+  int16_t nwaitnotempty;      /* Number tasks waiting for not empty */
+};
+
 /* This structure defines a message queue */
 
 struct mqueue_inode_s
 {
+  struct mqueue_cmn_s cmn;    /* Common prologue */
```
Comment (pkarashchenko) [inline include/nuttx/mqueue.h:109]: Can we name it something like `struct mqueue_wait_node_s wnode;`? This is just a suggestion.

File: sched/mqueue/msg.h [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+EXTERN struct list_node g_msgfreelist;
+
+/********************************************************************************
+ * Public Function Prototypes
+ ********************************************************************************/
+
+void               nxmsg_initialize(void);
+int                nxmsg_alloc(FAR struct msgq_s **pmsgq);
+void               nxmsg_free(FAR struct msgq_s *msgq);
+FAR struct msgq_s *nxmsg_lookup(key_t key);
```
Comment (pkarashchenko) [inline sched/mqueue/msg.h:90]: description is missing

File: sched/mqueue/msg.h [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+EXTERN struct list_node g_msgfreelist;
+
+/********************************************************************************
```
Comment (pkarashchenko) [inline sched/mqueue/msg.h:83]: comments in this file seems to be too long

File: sched/mqueue/msgctl.c [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+ *   used with repeated MSG_STAT or MSG_STAT_ANY operations to obtain
+ *   information about all queues on the system.)  A successful
+ *   MSG_STAT or MSG_STAT_ANY operation returns the identifier of the
+ *   queue whose index was given in msqid.
+ *
+ *   On failure, -1 is returned and errno is set to indicate the error.
+ *
+ ****************************************************************************/
+
+int msgctl(int msqid, int cmd, struct msqid_ds *buf)
```
Comment (pkarashchenko) [inline sched/mqueue/msgctl.c:58]:
```suggestion
int msgctl(int msqid, int cmd, FAR struct msqid_ds *buf)
```

File: sched/mqueue/msginternal.c [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+      if (g_msgmgrs.msgqs[i] == NULL)
+        {
+          g_msgmgrs.msgqs[i] = msgq;
+          msgq->key = i + 1;
+          return msgq;
+        }
+    }
+
+  tmp = kmm_realloc(g_msgmgrs.msgqs, sizeof(FAR void *) *
+                                            (g_msgmgrs.nmsgq + MQ_PERBLOCK));
```
Comment (pkarashchenko) [inline sched/mqueue/msginternal.c:87]:
```suggestion
  tmp = kmm_realloc(g_msgmgrs.msgqs, sizeof(FAR void *) *
                    (g_msgmgrs.nmsgq + MQ_PERBLOCK));
```

File: include/nuttx/mqueue.h (hunk 10 above)
Comment (pkarashchenko) [reply include/nuttx/mqueue.h:109]: Ok. If that is already used then let's keep it

File: sched/mqueue/msg.h [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+ * Included Files
+ ********************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/compiler.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/list.h>
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msg.h:32]: Done

File: sched/mqueue/msginternal.c (hunk 14 above)
Comment (anchao) [reply sched/mqueue/msginternal.c:87]: Done

File: sched/mqueue/msgctl.c (hunk 13 above)
Comment (anchao) [reply sched/mqueue/msgctl.c:58]: Done

File: include/sys/msg.h (hunk 9 above)
Comment (anchao) [reply include/sys/msg.h:205]: Done

File: include/sys/msg.h (hunk 8 above)
Comment (anchao) [reply include/sys/msg.h:179]: Done

File: sched/mqueue/msg.h [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+EXTERN struct list_node g_msgfreelist;
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msg.h:81]: reference from other files:

sched/mqueue/msgrcv.c:219:  list_add_tail(&g_msgfreelist, &msg->node);
sched/mqueue/msgsnd.c:194:      msg = (FAR struct msgbuf_s *)list_remove_head(&g_msgfreelist);
sched/mqueue/msginternal.c:54:struct list_node g_msgfreelist = LIST_INITIAL_VALUE(g_msgfreelist);
sched/mqueue/msginternal.c:127:          list_add_tail(&g_msgfreelist, &msg->node);
sched/mqueue/msginternal.c:179:      list_add_tail(&g_msgfreelist, &entry->node);
sched/mqueue/msg.h:79:EXTERN struct list_node g_msgfreelist;

File: sched/mqueue/msginternal.c [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+
+  kmm_free(g_msgmgrs.msgqs[index]);
+  g_msgmgrs.msgqs[index] = NULL;
+}
+
+/****************************************************************************
+ * Name: nxmsg_lookup
+ ****************************************************************************/
+
+FAR struct msgq_s *nxmsg_lookup(key_t key)
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msginternal.c:190]: Done

File: sched/mqueue/msgctl.c [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+{
+  FAR struct msgq_s *msgq;
+  irqstate_t flags;
+  int ret = OK;
+
+  flags = enter_critical_section();
+
+  msgq = nxmsg_lookup((key_t)msqid);
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msgctl.c:66]: Done

File: sched/mqueue/msginternal.c [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+{
+  uint8_t             nmsgq;  /* The number of groups of msgs array */
+  FAR struct msgq_s **msgqs;  /* The pointer of two layer file descriptors array */
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct nxmsg_manager_s g_msgmgrs;
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msginternal.c:48]: Done

File: sched/mqueue/msgsnd.c [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+      /* No.. Not in an interrupt handler.  Is the message queue FULL? */
+
+      if (msgq->nmsgs >= msgq->maxmsgs)
+        {
+          /* Yes.. the message queue is full.  Wait for space to become
+           * available in the message queue.
+           */
+
+          ret = msgsnd_wait(msgq, msgflg);
+        }
```
[the remark this answers is not in the document]
Comment (anchao) [reply sched/mqueue/msgsnd.c:187]: Done

File: include/nuttx/mqueue.h (hunk 10 above)
Comment (anchao) [reply include/nuttx/mqueue.h:109]: Done

File: sched/mqueue/msg.h (hunk 11 above)
Comment (anchao) [reply sched/mqueue/msg.h:90]: Done

File: sched/mqueue/msgget.c [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+ *   msgflg - Operations and permissions flag
+ *
+ * Returned Value:
+ *   On success, msgget() returns the message queue identifier (a
+ *   nonnegative integer).  On failure, -1 is returned, and errno is
+ *   set to indicate the error.
+ *
+ ****************************************************************************/
+
+int msgget(key_t key, int msgflg)
```
Comment (pkarashchenko) [inline sched/mqueue/msgget.c:53]: Do we need to use `key_t` or `int` for key?

File: sched/mqueue/msgrcv.c [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+          if (msgtyp < 0)
+            {
+              if (newmsg == NULL || ((FAR struct msgbuf_s *)newmsg)->mtype >
+                                    ((FAR struct msgbuf_s *)node)->mtype)
+                {
+                  newmsg = node;
+                }
+            }
+          else if (msgtyp == 0 ||
+                   (msgflg & MSG_EXCEPT) != 0 ||
```
Comment (xiaoxiang781216) [inline sched/mqueue/msgrcv.c:92]: the logic is wrong?

File: sched/mqueue/msgrcv.c [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+
+      return -ECANCELED;
+    }
+#endif
+
+  /* Get the message from the head of the queue */
+
+  while (1)
+    {
+      list_for_every(&msgq->msglist, node)
```
Comment (xiaoxiang781216) [inline sched/mqueue/msgrcv.c:66]: list_for_every_entry to avoid the cast at line 85-86

File: sched/mqueue/msgrcv.c [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+
+/****************************************************************************
+ * Name: msgrcv_wait
+ ****************************************************************************/
+
+static int msgrcv_wait(FAR struct msgq_s *msgq, FAR struct msgbuf_s **rcvmsg,
+                       long msgtyp, int msgflg)
+{
+  FAR struct list_node *newmsg = NULL;
+  FAR struct list_node *node;
```
Comment (xiaoxiang781216) [inline sched/mqueue/msgrcv.c:44]: change the type of newmsg and node to msgbuf_s

File: sched/mqueue/msgrcv.c [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+
+      rtcb          = this_task();
+      rtcb->waitobj = msgq;
+      msgq->cmn.nwaitnotempty++;
+
+      /* Initialize the 'errcode" used to communication wake-up error
+       * conditions.
+       */
+
+      rtcb->errcode  = OK;
```
Comment (xiaoxiang781216) [inline sched/mqueue/msgrcv.c:123]: remove the extra space before =

File: sched/mqueue/msgrcv.c [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+      goto errout_with_critical;
+    }
+
+  ret = msgrcv_wait(msgq, &msg, msgtyp, msgflg);
+  if (ret < 0)
+    {
+      goto errout_with_critical;
+    }
+
+  msgq->nmsgs--;
```
Comment (xiaoxiang781216) [inline sched/mqueue/msgrcv.c:215]: move to msgrcv_wait

File: sched/mqueue/msgget.c (hunk 21 above)
Comment (anchao) [reply sched/mqueue/msgget.c:53]: Done

File: sched/mqueue/msgrcv.c (hunk 22 above)
Comment (anchao) [reply sched/mqueue/msgrcv.c:92]: Done

File: sched/mqueue/msgrcv.c (hunk 23 above)
Comment (anchao) [reply sched/mqueue/msgrcv.c:66]: Done

File: sched/mqueue/msgrcv.c (hunk 25 above)
Comment (anchao) [reply sched/mqueue/msgrcv.c:123]: Done

File: sched/mqueue/msgrcv.c (hunk 26 above)
Comment (anchao) [reply sched/mqueue/msgrcv.c:215]: Done

File: sched/mqueue/msgrcv.c (hunk 24 above)
Comment (anchao) [reply sched/mqueue/msgrcv.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**